Device plugin failures: KEP docs (#47029)
kubernetes/website: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md

`/var/lib/kubelet/device-plugins` when it starts. A device plugin can monitor the deletion
of its Unix socket and re-register itself upon such an event.

### Device plugin and unhealthy devices

There are cases when devices fail or being shut down. Device Plugin responsibility

> in this case is to notify kubelet about this situation using the `ListAndWatchResponse` API.

Reviewer: I'd not be so specific in this form of documentation, I'd have generecally mentioned the device plugin interface and linked to it, but this is my personal taste and being specific is not wrong either IMO.

Author: I wanted to do more, but tight deadline is here. I will try to spend time later on further improvements

> in this case is to notify kubelet about this situation using the `ListAndWatchResponse` stream.

Reviewer 2: What do you mean by stream in this case?

If that is the GRPC call, I would maybe just drop "stream".
```suggestion
in this case is to notify kubelet about this situation using the `ListAndWatchResponse` API.
```

Once device is marked as unhealthy, kubelet will decrease the allocatable count
for this resource on the Node to reflect how many devices can be used for scheduling new pods.
Capacity count for the resource will not change.

Pods that were assigned to the failed devices will continue be assigned to this device.
It is typical that code relying on the device will start failing and Pod may get
into Failed phase if `restartPolicy` for the Pod was not `Always` or enter the crash loop
otherwise.

> Before the Kuberenetes version 1.31, the way to know whether Pod is associated with the

Reviewer 2: ```suggestion
Before the Kubernetes version 1.31, the way to know whether Pod is associated with the
```

> Before the Kubernetes version 1.31, the way to know whether Pod is associated with the
> failed device was to use the [PodResources API](#monitoring-device-plugin-resources).

Reviewer 3: ```suggestion
Before Kubernetes v1.31, the way to know whether or not a Pod is associated with the
failed device is to use the [PodResources API](#monitoring-device-plugin-resources).
```

> {{< feature-state for_k8s_version="v1.31" state="alpha" >}}

Reviewer 4: Consider using the "feature_gate_name" option for `feature-state` shortcode. This will make it more data-driven by pulling from the feature gate description file.

```suggestion
{{< feature-state feature_gate_name="ResourceHealthStatus" >}}
```

By enabling the feature gate `ResourceHealthStatus`, the new field `AllocatedResourcesStatus`
will be added to each ContainerStatus in the pod status object that will contain
Health information for each device assigned to the container.

This status can be used to understand whether pod failure or misbehavior was associated

> with the device failure.

Reviewer 5: Does that example look correct?

> with the device failure.

Reviewer 5: (would like confirmation I guessed right)

> with the device failure.

Reviewer: The device plugin API doesn't have provision to report unhealthy reason. The device will thus report to the kubelet just healthy/unhealthy. The current user-facing API (in the pod object) also report just healthy/unhealty. So the example is realistic, but in the pod status the user won't actually find "unhealthy because overheating" in 1.31, but just "unhealthy". @SergeyKanzhelev to keep me honest here.

> with the device failure.

Reviewer 5: OK cool. Polishing this up post-merge is fine.

How do we feel about a separate controller annotating the Pod with more detail vs. reporting an Event vs. neither of those?

> with the device failure.

Reviewer: @sftim IMO this is great input for the next stages of this work!
